{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7055f9ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Change the input OR to BETA (in smoke and drink) for EAS population\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "eas_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/CAS/\"\n",
    "binary_trait = ['smoke', 'drink']\n",
    "\n",
    "for trait in binary_trait:\n",
    "    for group in range(1, 11):\n",
    "        print(f\"Processing group {group}, trait {trait}\")\n",
    "        trait_file = os.path.join(eas_base_dir, f\"{trait}/group_{group}/gwas/train.Pheno.glm.logistic\")\n",
    "        output_file = os.path.join(eas_base_dir, f\"{trait}/group_{group}/gwas/train.Pheno.glm.linear\")\n",
    "        if not os.path.exists(trait_file):\n",
    "            print(f\"File {trait_file} does not exist. Skipping.\")\n",
    "            continue\n",
    "        df = pd.read_csv(trait_file, sep=\"\\t\")\n",
    "        df['BETA'] = df['OR'].apply(lambda x: 0 if x <= 0 else np.log(x))\n",
    "        df.to_csv(output_file, sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99f4f17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine tune and test to construct a valid dataset for EAS\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# demo gwas: /data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/CAS/alt/group_1/gwas/train.Pheno.glm.linear\n",
    "trait_list = ['waist', 'height', 'pulse', 'dbp', 'sbp', 'smoke', 'drink', 'bmi', 'wbc', 'rbc', 'hb', 'plt', 'lymph', 'mono', 'neut', 'eos', 'alt', 'ast', 'bun', 'cholesterol', 'creatinine', 'glucose', 'ggt', 'hdl', 'ldl', 'triglycerides', 'ua']\n",
    "eas_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/CAS/\"\n",
    "output_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/\"\n",
    "eas_bfile_path = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/CAS/geno/CAS_final/CAS_merged_qc_final\"\n",
    "\n",
    "for trait in trait_list:\n",
    "    for group in range(1, 11):\n",
    "        print(f\"Processing group {group}, trait {trait}\")\n",
    "        # tune_id_path = os.path.join(eas_base_dir, f\"{trait}/group_{group}/ids/tune_ids.txt\")\n",
    "        # test_id_path = os.path.join(eas_base_dir, f\"{trait}/group_{group}/ids/test_ids.txt\")\n",
    "        combine_id_path = os.path.join(eas_base_dir, f\"{trait}/group_{group}/ids/combined_ids.txt\")\n",
    "        # tune_output_dir = os.path.join(output_base_dir, f\"tune/{trait}/group_{group}/\")\n",
    "        test_output_dir = os.path.join(output_base_dir, f\"test/{trait}/group_{group}/\")\n",
    "        # os.makedirs(tune_output_dir, exist_ok=True)\n",
    "        os.makedirs(test_output_dir, exist_ok=True)\n",
    "        # tune_output_prefix = os.path.join(tune_output_dir, \"tune\")\n",
    "        # test_output_prefix = os.path.join(test_output_dir, \"test\")\n",
    "        combine_output_prefix = os.path.join(test_output_dir, \"combine\")\n",
    "\n",
    "        # tune_command = f\"plink2 --bfile {eas_bfile_path} --keep {tune_id_path} --make-bed --out {tune_output_prefix}\"\n",
    "        # test_command = f\"plink2 --bfile {eas_bfile_path} --keep {test_id_path} --make-bed --out {test_output_prefix}\"\n",
    "        combine_command = f\"plink2 --bfile {eas_bfile_path} --keep {combine_id_path} --make-bed --out {combine_output_prefix}\"\n",
    "        # os.system(tune_command)\n",
    "        # os.system(test_command)\n",
    "        os.system(combine_command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c6a5cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run C+T with different r2 and window size for EAS\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "trait_list = ['waist', 'height', 'pulse', 'dbp', 'sbp', 'smoke', 'drink', 'bmi', 'wbc', 'rbc', 'hb', 'plt', 'lymph', 'mono', 'neut', 'eos', 'alt', 'ast', 'bun', 'cholesterol', 'creatinine', 'glucose', 'ggt', 'hdl', 'ldl', 'triglycerides', 'ua']\n",
    "r2_list = [0.1]\n",
    "windows_list = [500]\n",
    "\n",
    "eas_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/CAS/\"\n",
    "ct_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/\"\n",
    "output_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/clumped/\"\n",
    "\n",
    "for trait in trait_list:\n",
    "    for group in range(1, 11):\n",
    "        print(f\"Processing group {group}, trait {trait}\")\n",
    "        gwas_file = os.path.join(eas_base_dir, f\"{trait}/group_{group}/gwas/train.Pheno.glm.linear\")\n",
    "        bfile_path = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/tlprs/reference/EAS_1kg/1000G.EAS.QC.hm3.ind\"\n",
    "        if not os.path.exists(gwas_file):\n",
    "            print(f\"File {gwas_file} does not exist. Skipping.\")\n",
    "            continue\n",
    "        for r2 in r2_list:\n",
    "            for window in windows_list:\n",
    "                output_dir = os.path.join(output_base_dir, f\"{trait}/group_{group}/r2_{r2}_w_{window}/\")\n",
    "                os.makedirs(output_dir, exist_ok=True)\n",
    "                output_prefix = os.path.join(output_dir, \"res\")\n",
    "                clump_command = f\"plink2 --bfile {bfile_path} --clump {gwas_file} --clump-p1 1 --clump-r2 {r2} --clump-kb {window} --clump-snp-field ID --clump-field P --out {output_prefix}\"\n",
    "                os.system(clump_command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e404c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Change the input OR to BETA (in smoke and drink)\n",
    "# /data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/UKB_EUR/train/gwas/fold_1/gwas_chr14_binary.p20116_int.glm.logistic\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "eur_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/UKB_EUR/train/gwas/\"\n",
    "binary_trait_dict = {\n",
    "    'p20116': 'smoke',\n",
    "    'p20117': 'drink'\n",
    "}\n",
    "\n",
    "for trait_id, trait in binary_trait_dict.items():\n",
    "    for group in range(1, 11):\n",
    "        print(f\"Processing group {group}, trait {trait}\")\n",
    "        for chrom in range(1, 23):\n",
    "            print(f\"Processing chromosome {chrom}\")\n",
    "            trait_file = os.path.join(eur_base_dir, f\"fold_{group}/gwas_chr{chrom}_binary.{trait_id}_int.glm.logistic\")\n",
    "            output_file = os.path.join(eur_base_dir, f\"fold_{group}/gwas_chr{chrom}_binary.{trait_id}_int.glm.linear\")\n",
    "            if not os.path.exists(trait_file):\n",
    "                print(f\"File {trait_file} does not exist. Skipping.\")\n",
    "                continue\n",
    "            df = pd.read_csv(trait_file, sep=\"\\t\")\n",
    "            df['BETA'] = df['OR'].apply(lambda x: 0 if x <= 0 else np.log(x))\n",
    "            df.to_csv(output_file, sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c81ce8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run C+T with different r2 and window size\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "trait_dict = {\n",
    "    'p48': 'waist',\n",
    "    'p50': 'height',\n",
    "    'p102': 'pulse',\n",
    "    'p4079': 'dbp',\n",
    "    'p4080': 'sbp',\n",
    "    'p20116': 'smoke',\n",
    "    'p20117': 'drink',\n",
    "    'p21001': 'bmi',\n",
    "    'p30000': 'wbc',\n",
    "    'p30010': 'rbc',\n",
    "    'p30020':'hb',\n",
    "    'p30080': 'plt',\n",
    "    'p30120': 'lymph',\n",
    "    'p30130': 'mono',\n",
    "    'p30140': 'neut',\n",
    "    'p30150': 'eos',\n",
    "    'p30620': 'alt',\n",
    "    'p30650': 'ast',\n",
    "    'p30670': 'bun',\n",
    "    'p30690': 'cholesterol',\n",
    "    'p30700': 'creatinine',\n",
    "    'p30730': 'ggt',\n",
    "    'p30740': 'glucose',\n",
    "    'p30760': 'hdl',\n",
    "    'p30780': 'ldl',\n",
    "    'p30870': 'triglycerides',\n",
    "    'p30880': 'ua'\n",
    "}\n",
    "\n",
    "r2_list = [0.1]\n",
    "windows_list = [500]\n",
    "\n",
    "eur_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/UKB_EUR/train/gwas/\"\n",
    "eur_tune_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/Cross_Validation/UKB_EUR/tune/\"\n",
    "output_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/clumped/\"\n",
    "\n",
    "for trait_id, trait in trait_dict.items():\n",
    "    for group in range(10, 11):\n",
    "        print(f\"Processing group {group}, trait {trait}\")\n",
    "        for chrom in range(1, 23):\n",
    "            gwas_file = os.path.join(eur_base_dir, f\"fold_{group}/gwas_chr{chrom}.{trait_id}_int.glm.linear\")\n",
    "            if trait_id in ['p20116', 'p20117']:\n",
    "                gwas_file = os.path.join(eur_base_dir, f\"fold_{group}/gwas_chr{chrom}_binary.{trait_id}_int.glm.linear\")\n",
    "            if not os.path.exists(gwas_file):\n",
    "                print(f\"File {gwas_file} does not exist. Skipping.\")\n",
    "                continue\n",
    "            tune_bfile = os.path.join(eur_tune_dir, f\"fold_{group}/chr{chrom}\")\n",
    "            for r2 in r2_list:\n",
    "                for window in windows_list:\n",
    "                    output_dir = os.path.join(output_base_dir, f\"{trait}/group_{group}/r2_{r2}_w_{window}/\")\n",
    "                    os.makedirs(output_dir, exist_ok=True)\n",
    "                    output_prefix = os.path.join(output_dir, f\"ukb_chr{chrom}\")\n",
    "                    clump_command = f\"plink2 --bfile {tune_bfile} --rm-dup exclude-mismatch --clump {gwas_file} --clump-p1 1 --clump-r2 {r2} --clump-kb {window} --clump-snp-field ID --clump-field P --threads 20 --out {output_prefix}\"\n",
    "                    os.system(clump_command)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
