{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f961ae38",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "full_clump_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/full_model/EAS/\"\n",
    "full_eas_gwas_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/CAS/gwas/gwas/\"\n",
    "output_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/full_model/EAS/\"\n",
    "\n",
    "# trait_list = ['waist', 'height', 'pulse', 'dbp', 'sbp', 'smoke', 'drink', 'bmi', 'wbc', 'rbc', 'hb', 'plt', 'lymph', 'mono', 'neut', 'eos', 'alt', 'ast', 'bun', 'cholesterol', 'creatinine', 'glucose', 'ggt', 'hdl', 'ldl', 'triglycerides', 'ua']\n",
    "# trait_list = ['alt', 'ast', 'bun', 'cholesterol', 'creatinine', 'glucose', 'ggt', 'hdl', 'ldl', 'triglycerides', 'ua']\n",
    "trait_list = ['alt']\n",
    "\n",
    "for trait in trait_list:\n",
    "    print(f\"Processing trait: {trait}\")\n",
    "    clump_file = os.path.join(full_clump_dir, f\"{trait}.clumps\")\n",
    "    gwas_path = os.path.join(full_eas_gwas_dir, f\"{trait}_int.{trait}_int.glm.linear\")\n",
    "    if trait in ['smoke', 'drink']:\n",
    "        gwas_path = os.path.join(full_eas_gwas_dir, f\"{trait}_raw.{trait}_raw.glm.logistic\")\n",
    "    df = pd.read_csv(clump_file, sep=\"\\t\")\n",
    "    df_filtered = df[[\"#CHROM\", \"POS\", \"ID\"]]\n",
    "\n",
    "    gwas = pd.read_csv(gwas_path, sep=\"\\t\")\n",
    "    # if OR column exists, change OR into beta\n",
    "    if 'OR' in gwas.columns:\n",
    "        gwas['BETA'] = np.log(gwas['OR'])\n",
    "    gwas = gwas[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"BETA\", \"P\"]]\n",
    "    merged = pd.merge(df_filtered, gwas, on=[\"#CHROM\", \"POS\", \"ID\"], how=\"inner\")\n",
    "    if merged.shape[0] != df_filtered.shape[0]:\n",
    "        print(f\"Warning: Mismatch in SNP counts for trait {trait}. Clumped SNPs: {df_filtered.shape[0]}, Merged SNPs: {merged.shape[0]}\")\n",
    "    output_path = os.path.join(output_dir, f\"{trait}.tsv\")\n",
    "    merged.to_csv(output_path, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dc5eeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate PRS based on the filtered clump files on EAS test data\n",
    "# test data demo: /data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/test/alt/group_1/test\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "# trait_list = ['waist', 'height', 'pulse', 'dbp', 'sbp', 'smoke', 'drink', 'bmi', 'wbc', 'rbc', 'hb', 'plt', 'lymph', 'mono', 'neut', 'eos', 'alt', 'ast', 'bun', 'cholesterol', 'creatinine', 'glucose', 'ggt', 'hdl', 'ldl', 'triglycerides', 'ua']\n",
    "trait_list = ['alt']\n",
    "clump_res_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/full_model/EAS/\"\n",
    "test_bfile_path = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/UKB/geno/Chinese/1_merged/merged\"\n",
    "output_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/test_out_sample/\"\n",
    "\n",
    "for trait in trait_list:\n",
    "    print(f\"Processing trait {trait}\")\n",
    "    clump_file = os.path.join(clump_res_dir, f\"{trait}.tsv\")\n",
    "    if not os.path.exists(clump_file):\n",
    "        print(f\"File {clump_file} does not exist. Skipping.\")\n",
    "        continue\n",
    "    output_subdir = os.path.join(output_dir, f\"EAS/\")\n",
    "    os.makedirs(output_subdir, exist_ok=True)\n",
    "    output_prefix = os.path.join(output_subdir, f\"{trait}\")\n",
    "    prs_command = f\"plink2 --bfile {test_bfile_path} --score {clump_file} 3 5 6 header no-mean-imputation --out {output_prefix}\"\n",
    "    os.system(prs_command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a584680",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "full_clump_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/full_model/EUR/\"\n",
    "full_eur_gwas_path = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/UKB/gwas/White_British/\"\n",
    "output_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/full_model/EUR/\"\n",
    "\n",
    "trait_dict = {\n",
    "        'p48': 'waist',\n",
    "        'p50': 'height',\n",
    "        'p102': 'pulse',\n",
    "        'p4079': 'dbp',\n",
    "        'p4080': 'sbp',\n",
    "        'p20116': 'smoke',\n",
    "        'p20117': 'drink',\n",
    "        'p21001': 'bmi',\n",
    "        'p30000': 'wbc',\n",
    "        'p30010': 'rbc',\n",
    "        'p30020':'hb',\n",
    "        'p30080': 'plt',\n",
    "        'p30120': 'lymph',\n",
    "        'p30130': 'mono',\n",
    "        'p30140': 'neut',\n",
    "        'p30150': 'eos',\n",
    "        'p30620': 'alt',\n",
    "        'p30650': 'ast',\n",
    "        'p30670': 'bun',\n",
    "        'p30690': 'cholesterol',\n",
    "        'p30700': 'creatinine',\n",
    "        'p30730': 'ggt',\n",
    "        'p30740': 'glucose',\n",
    "        'p30760': 'hdl',\n",
    "        'p30780': 'ldl',\n",
    "        'p30870': 'triglycerides',\n",
    "        'p30880': 'ua'\n",
    "    }\n",
    "\n",
    "for trait_id, trait_name in trait_dict.items():\n",
    "    res_df = []\n",
    "    for chrom in range(1, 23):\n",
    "        print(f\"Processing trait: {trait_name} of chromosome chr{chrom}\")\n",
    "        clump_file = os.path.join(full_clump_dir, f\"{trait_name}/chr_{chrom}.clumps\")\n",
    "        if not os.path.exists(clump_file):\n",
    "            print(f\"File {clump_file} does not exist. Skipping.\")\n",
    "            continue\n",
    "        gwas_path = os.path.join(full_eur_gwas_path, f\"{trait_id}_int_chr{chrom}.{trait_id}_int.glm.linear\")\n",
    "        if trait_name in ['smoke', 'drink']:\n",
    "            gwas_path = os.path.join(full_eur_gwas_path, f\"{trait_id}_int_chr{chrom}.{trait_id}_int.glm.logistic\")\n",
    "        df = pd.read_csv(clump_file, sep=\"\\t\")\n",
    "        df_filtered = df[[\"#CHROM\", \"POS\", \"ID\"]]\n",
    "        gwas = pd.read_csv(gwas_path, sep=\"\\t\")\n",
    "        # if OR column exists, change OR into beta\n",
    "        if 'OR' in gwas.columns:\n",
    "            gwas['BETA'] = np.log(gwas['OR'])\n",
    "        gwas = gwas[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"BETA\", \"P\"]]\n",
    "        merged = pd.merge(df_filtered, gwas, on=[\"#CHROM\", \"POS\", \"ID\"], how=\"inner\")\n",
    "        if merged.shape[0] != df_filtered.shape[0]:\n",
    "            print(f\"Warning: Mismatch in SNP counts for trait {trait_name} chr{chrom}. Clumped SNPs: {df_filtered.shape[0]}, Merged SNPs: {merged.shape[0]}\")\n",
    "        # output_path = os.path.join(output_dir, f\"{trait_name}/chr{chrom}.tsv\")\n",
    "        # merged.to_csv(output_path, sep=\"\\t\", index=False)\n",
    "        res_df.append(merged)\n",
    "    if len(res_df) > 0:\n",
    "        final_df = pd.concat(res_df, ignore_index=True)\n",
    "        final_output_path = os.path.join(output_dir, f\"{trait_name}/{trait_name}.tsv\")\n",
    "        final_df.to_csv(final_output_path, sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71670058",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate PRS based on the filtered clump files on EAS test data\n",
    "# test data demo: /data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/test/alt/group_1/test\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "trait_list = ['waist', 'height', 'pulse', 'dbp', 'sbp', 'smoke', 'drink', 'bmi', 'wbc', 'rbc', 'hb', 'plt', 'lymph', 'mono', 'neut', 'eos', 'alt', 'ast', 'bun', 'cholesterol', 'creatinine', 'glucose', 'ggt', 'hdl', 'ldl', 'triglycerides', 'ua']\n",
    "# trait_list = ['alt']\n",
    "clump_res_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/full_model/EUR/\"\n",
    "test_bfile_path = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/UKB/geno/Chinese/1_merged/merged\"\n",
    "output_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/ct/res/test_out_sample/\"\n",
    "\n",
    "for trait in trait_list:\n",
    "    print(f\"Processing trait {trait}\")\n",
    "    clump_file = os.path.join(clump_res_dir, f\"{trait}/{trait}.tsv\")\n",
    "    if not os.path.exists(clump_file):\n",
    "        print(f\"File {clump_file} does not exist. Skipping.\")\n",
    "        continue\n",
    "    output_subdir = os.path.join(output_dir, f\"EUR/\")\n",
    "    os.makedirs(output_subdir, exist_ok=True)\n",
    "    output_prefix = os.path.join(output_subdir, f\"{trait}\")\n",
    "    prs_command = f\"plink2 --bfile {test_bfile_path} --score {clump_file} 3 5 6 header no-mean-imputation --out {output_prefix}\"\n",
    "    os.system(prs_command)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
