{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5054592",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "def process_single_file(gwas_file, output_path):\n",
    "    \"\"\"\n",
    "    Process a single GWAS file to extract and save the required data.\n",
    "    \n",
    "    Parameters:\n",
    "    gwas_file (str): Path to the GWAS file.\n",
    "    output_path (str): Directory where the output file will be saved.\n",
    "    \n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    if gwas_file.endswith('.merged.glm.linear'):\n",
    "        if \"BETA\" in pd.read_csv(gwas_file, sep='\\t', nrows=1).columns:\n",
    "            df = pd.read_csv(gwas_file, sep='\\t', usecols=['ID', 'ALT', 'REF', 'A1_FREQ', 'BETA', 'SE', 'P', 'OBS_CT'])\n",
    "            df = df[['ID', 'ALT', 'REF', 'A1_FREQ', 'BETA', 'SE', 'P', 'OBS_CT']]\n",
    "        else:\n",
    "            df = pd.read_csv(gwas_file, sep='\\t', usecols=['ID', 'ALT', 'REF', 'A1_FREQ', 'OR', 'LOG(OR)_SE', 'P', 'OBS_CT'])\n",
    "            df[\"BETA\"] = np.log(df[\"OR\"])\n",
    "            df[\"SE\"] = df[\"LOG(OR)_SE\"]\n",
    "            df = df[['ID', 'ALT', 'REF', 'A1_FREQ', 'BETA', 'SE', 'P', 'OBS_CT']]\n",
    "        df.columns = ['SNP', 'A1', 'A2', 'freq', 'b', 'se', 'p', 'N']\n",
    "        # df['p'] = df['p'].clip(1e-100, 1)\n",
    "        df['N'] = df['N'].fillna(0).astype(int)\n",
    "        df['b'] = df['b'].astype(float)\n",
    "        df['se'] = df['se'].astype(float)\n",
    "        df['freq'] = df['freq'].astype(float)\n",
    "        df['SNP'] = df['SNP'].astype(str)\n",
    "        df = df.drop_duplicates(subset=['SNP'])\n",
    "        df = df[df[\"p\"] > 1e-100]\n",
    "        df = df[df[\"freq\"] > 0.01]\n",
    "        # draw the distribution of se\n",
    "        df = df[df['se'] > 0]\n",
    "        df = df.dropna(subset=['b', 'se', 'p'])\n",
    "        df.to_csv(output_path, sep='\\t', index=False)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    gwas_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/UKB/merged_gwas/White_British/gwas/\"\n",
    "    output_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/polypred/train/EUR/4_sbayesr/\"\n",
    "    trait_dict = {\n",
    "        # 'p48': 'waist',\n",
    "        'p50': 'height',\n",
    "        'p102': 'pulse',\n",
    "        'p4079': 'dbp',\n",
    "        'p4080': 'sbp',\n",
    "        'p20116': 'smoke',\n",
    "        'p20117': 'drink',\n",
    "        'p21001': 'bmi',\n",
    "        # 'p21002': 'weight',\n",
    "        'p30000': 'wbc',\n",
    "        'p30010': 'rbc',\n",
    "        'p30020':'hb',\n",
    "        'p30080': 'plt',\n",
    "        'p30120': 'lymph',\n",
    "        'p30130': 'mono',\n",
    "        'p30140': 'neut',\n",
    "        'p30150': 'eos',\n",
    "        'p30620': 'alt',\n",
    "        'p30650': 'ast',\n",
    "        'p30670': 'bun',\n",
    "        'p30690': 'cholesterol',\n",
    "        'p30700': 'creatinine',\n",
    "        'p30730': 'ggt',\n",
    "        'p30740': 'glucose',\n",
    "        'p30760': 'hdl',\n",
    "        'p30780': 'ldl',\n",
    "        'p30870': 'triglycerides',\n",
    "        'p30880': 'ua'\n",
    "    }\n",
    "\n",
    "    for trait, trait_name in trait_dict.items():\n",
    "        gwas_file = os.path.join(gwas_base_dir, f\"{trait}_int.merged.glm.linear\")\n",
    "        output_path = os.path.join(output_base_dir, f\"{trait_name}.ma\")\n",
    "        \n",
    "        if os.path.exists(gwas_file):\n",
    "            process_single_file(gwas_file, output_path)\n",
    "            print(f\"Processed {gwas_file} and saved to {output_path}\")\n",
    "        else:\n",
    "            print(f\"File {gwas_file} does not exist.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d8ae60",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "def process_single_file(gwas_file, output_path):\n",
    "    \"\"\"\n",
    "    Process a single GWAS file to extract and save the required data.\n",
    "    \n",
    "    Parameters:\n",
    "    gwas_file (str): Path to the GWAS file.\n",
    "    output_path (str): Directory where the output file will be saved.\n",
    "    \n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    if gwas_file.endswith('.merged.glm.linear'):\n",
    "        if \"BETA\" in pd.read_csv(gwas_file, sep='\\t', nrows=1).columns:\n",
    "            df = pd.read_csv(gwas_file, sep='\\t', usecols=['#CHROM', 'POS', 'ID', 'ALT', 'REF', 'A1_FREQ', 'BETA', 'SE', 'P', 'OBS_CT'])\n",
    "            df = df[['#CHROM', 'POS', 'ID', 'ALT', 'REF', 'A1_FREQ', 'BETA', 'SE', 'P', 'OBS_CT']]\n",
    "        else:\n",
    "            df = pd.read_csv(gwas_file, sep='\\t', usecols=['#CHROM', 'POS', 'ID', 'ALT', 'REF', 'A1_FREQ', 'OR', 'LOG(OR)_SE', 'P', 'OBS_CT'])\n",
    "            df[\"BETA\"] = np.log(df[\"OR\"])\n",
    "            df[\"SE\"] = df[\"LOG(OR)_SE\"]\n",
    "            df = df[['#CHROM', 'POS', 'ID', 'ALT', 'REF', 'A1_FREQ', 'BETA', 'SE', 'P', 'OBS_CT']]\n",
    "        df.columns = ['CHR', 'POS', 'SNP', 'A1', 'A2', 'freq', 'b', 'se', 'p', 'N']\n",
    "        # df['p'] = df['p'].clip(1e-100, 1)\n",
    "        df['N'] = df['N'].fillna(0).astype(int)\n",
    "        df['b'] = df['b'].astype(float)\n",
    "        df['se'] = df['se'].astype(float)\n",
    "        df['freq'] = df['freq'].astype(float)\n",
    "        df['SNP'] = df['SNP'].astype(str)\n",
    "        df = df.drop_duplicates(subset=['SNP'])\n",
    "        # remove the variants in 8477797 - 33448354 in chromosome 6\n",
    "        df = df[(df['CHR'] != '6') | ((df['POS'] < 8477797) | (df['POS'] > 33448354))]\n",
    "        df = df[df[\"p\"] > 1e-100]\n",
    "        df = df[df[\"freq\"] > 0.01]\n",
    "        # draw the distribution of se\n",
    "        df = df[df['se'] > 0]\n",
    "        df = df.dropna(subset=['b', 'se', 'p'])\n",
    "        res = df[['SNP', 'A1', 'A2', 'freq', 'b', 'se', 'p', 'N']]\n",
    "        res.to_csv(output_path, sep='\\t', index=False)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    gwas_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/real_data/UKB/merged_gwas/White_British/gwas/\"\n",
    "    output_base_dir = \"/data1/jiapl_group/lishuhua/project/PRS_benchmark/software/polypred/train/EUR/4_sbayesr/\"\n",
    "    trait_dict = {\n",
    "        # 'p48': 'waist',\n",
    "        'p50': 'height',\n",
    "        # 'p102': 'pulse',\n",
    "        # 'p4079': 'dbp',\n",
    "        # 'p4080': 'sbp',\n",
    "        # 'p20116': 'smoke',\n",
    "        # 'p20117': 'drink',\n",
    "        # 'p21001': 'bmi',\n",
    "        # 'p21002': 'weight',\n",
    "        # 'p30000': 'wbc',\n",
    "        # 'p30010': 'rbc',\n",
    "        # 'p30020':'hb',\n",
    "        # 'p30080': 'plt',\n",
    "        # 'p30120': 'lymph',\n",
    "        # 'p30130': 'mono',\n",
    "        # 'p30140': 'neut',\n",
    "        # 'p30150': 'eos',\n",
    "        # 'p30620': 'alt',\n",
    "        # 'p30650': 'ast',\n",
    "        # 'p30670': 'bun',\n",
    "        # 'p30690': 'cholesterol',\n",
    "        # 'p30700': 'creatinine',\n",
    "        # 'p30730': 'ggt',\n",
    "        # 'p30740': 'glucose',\n",
    "        # 'p30760': 'hdl',\n",
    "        # 'p30780': 'ldl',\n",
    "        # 'p30870': 'triglycerides',\n",
    "        'p30880': 'ua'\n",
    "    }\n",
    "\n",
    "    for trait, trait_name in trait_dict.items():\n",
    "        gwas_file = os.path.join(gwas_base_dir, f\"{trait}_int.merged.glm.linear\")\n",
    "        output_path = os.path.join(output_base_dir, f\"{trait_name}.ma\")\n",
    "        \n",
    "        if os.path.exists(gwas_file):\n",
    "            process_single_file(gwas_file, output_path)\n",
    "            print(f\"Processed {gwas_file} and saved to {output_path}\")\n",
    "        else:\n",
    "            print(f\"File {gwas_file} does not exist.\")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
